sonar-dotnet-shared-library/src/main/java/org/sonar/plugins/dotnet/tests/VisualStudioCoverageXmlReportParser.java

```diff
@@ -119,27 +131,67 @@ private void handleSourceFileTag(XmlParserHelper xmlParserHelper) {
         return;
       }
 
-      if (coveredLines.containsKey(id)) {
-        LOG.trace("Found covered lines for id '{}' for path '{}'", id, canonicalPath);
+      LineCoverage fileCoverage = lineCoveragePerFile.get(id);
+      fileCoverage.transferData(coverage, canonicalPath);
+    }
 
-        for (Integer line : coveredLines.get(id)) {
-          coverage.addHits(canonicalPath, line, 1);
-        }
-      }
+    private void checkRootTag(XmlParserHelper xmlParserHelper) {
+      xmlParserHelper.checkRootTag("results");
+    }
+  }
 
-      if (uncoveredLines.containsKey(id)) {
-        LOG.trace("Found uncovered lines for id '{}' for path '{}'", id, canonicalPath);
+  private class LineCoverage {
 
-        for (Integer line : uncoveredLines.get(id)) {
-          coverage.addHits(canonicalPath, line, 0);
-        }
-      }
+    int fileId;
+
+    // the key is the line ID
+    // the values are a list of booleans, where
+    // - 'True' means the line is covered or partially covered
+    // - 'False' means the line is not covered
+    private Map<Integer, List<Boolean>> coveragePerLine;
+
+    LineCoverage(int fileId) {
+      this.fileId = fileId;
+      this.coveragePerLine = new HashMap<>();
     }
 
-    private void checkRootTag(XmlParserHelper xmlParserHelper) {
-      xmlParserHelper.checkRootTag("results");
+    void addCoverageForLine(Integer lineId, Boolean isCovered) {
+      coveragePerLine.putIfAbsent(lineId, new ArrayList<>());
+      coveragePerLine.get(lineId).add(isCovered);
     }
 
+    /**
+     * Transfers the coverage data to the Coverage object.
+     * @param coverage - the object which centralizes coverage information.
+     * @param canonicalFilePath - the path of the file for which this class contains coverage data.
+     */
+    void transferData(Coverage coverage, String canonicalFilePath) {
+      if (!coveragePerLine.isEmpty()) {
```

Comment:
I was looking at the branch coverage report and not all the branches are covered. I think it's not possible to have this empty since we either add a entry or throw. WDYT?

Reply:
good point.
this is why full code coverage is important :)
I added a test with empty ranges and treating a NRE case

there would still be a case where the `source_files` is empty, but I don't think that's possible... but having an empty list of ranges for an existing file seems more plausible and I would like to be on the safe side